---
title: "RNA-seq from 4 cell populations"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

cd /hpc/users/hoffmg01/work/cmc_analysis/src
ml git pandoc
git pull
R

system("git pull")
rmarkdown::render("molec_profiling.Rmd", output_dir='./', intermediates_dir='./')


# http
# https://hoffmg01.u.hpc.mssm.edu/molec_profiling.html

--->


```{r load.packages, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages({
library(variancePartition)
library(limma)
library(edgeR)
library(mvIC)
library(data.table)
library(BiocParallel)
library(ggplot2)
library(gridExtra)
library(openxlsx)
})

# set.seed(1)
register(SnowParam(4))
# register(SerialParam())

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  cache.lazy = FALSE,  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```

```{r read.data}
file = "/sc/hydra/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/rnaseq/step1-merged/featureCount/featureCount.gene.tsv.gz"
df_featureCounts_gene = fread( file )

geneInfo = df_featureCounts_gene[,1:6]
geneInfo$EnsId = gsub("^(.*)(\\.\\S+)$", "\\1", geneInfo$Geneid)

geneCounts = df_featureCounts_gene[,7:ncol(df_featureCounts_gene)]
geneCounts = as.data.frame(geneCounts)
rownames(geneCounts) = geneInfo$Geneid

# identify samples with improper names
idx = grep("Undetermined", colnames(geneCounts))
message("Dropping samples with improper names:\n",  paste(colnames(geneCounts)[idx], collapse="\n"))
geneCounts = geneCounts[,-idx]

idx = which(colSums(geneCounts) < 1e5)
message("Dropping samples with zero read counts:\n",  paste(colnames(geneCounts)[idx], collapse="\n"))
geneCounts = geneCounts[,-idx]

# remove genes with duplicated EnsId
exclude = duplicated(geneInfo$EnsId)
geneInfo = geneInfo[!exclude,]
geneCounts = geneCounts[!exclude,]
rownames(geneCounts) = gsub("^(.*)(\\.\\S+)$", "\\1", geneInfo$Geneid)

```


```{r metadata}

# focused on this experiment
metadata1 = read.xlsx("/sc/hydra/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/meta-files/MolecularProfiling_MSSM_120samples_05.24.17-4.xlsx")

# form Synapse
metadataCMC = fread("/sc/hydra/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/meta-files/CMC_METADATA.csv")
```


```{r QC}

# extract information from sample names
info = data.frame( Name = colnames(geneCounts), stringsAsFactors=FALSE)
rownames(info) = info$Name
info$DissectionID = sapply(strsplit( info$Name, '\\.'), function(x) x[1])
info$CellType = sapply(strsplit( info$Name, '\\.'), function(x) x[2])
info$Assay = sapply(strsplit( info$Name, '\\.'), function(x) x[3])

# set color codes cor each cell type
colorCodes = c(
  GABA = '#66A61E',
  GLU = '#E6AB02',
  Olig = '#E7298A',
  MgAs = '#7570B3')

# set factor levels to be the same as the order of the colors
# this makes plotting easier
info$CellType = factor(info$CellType,names(colorCodes))

# add information about each sample
info = merge(info, metadata1, by.x="DissectionID", by.y='Dissection.ID')
# info = merge(info, metadataCMC, by.x='Individual.ID', by.y="Individual ID")
#  info[1:4, 1:20]

isexpr = rowSums(cpm(geneCounts)>1) >= 0.1*ncol(geneCounts)

dge = DGEList( geneCounts[isexpr,] )
dge = calcNormFactors(dge)
```


### PCA
```{r PCA}

geneExpr = cpm(dge, log=TRUE)

dcmp = prcomp( t(geneExpr), scale=TRUE)

df_pca = data.frame(dcmp$x)

df_pca = merge(df_pca, info, by.x="row.names", by.y="Name")

varFrac = dcmp$sdev^2 / sum(dcmp$sdev^2)

xlab = paste0('PC1 (', format(varFrac[1]*100, digits=3), '%)')
ylab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
ggplot(df_pca, aes(PC1, PC2, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(xlab) + ylab(ylab)

xlab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
ylab = paste0('PC3 (', format(varFrac[3]*100, digits=3), '%)')
ggplot(df_pca, aes(PC2, PC3, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(xlab) + ylab(ylab)
```      


### Differential expression
```{r dream}
vobj = voomWithDreamWeights(dge, ~ (1|CellType) + (1|Individual.ID) + Gender, info, plot=TRUE, BPPARAM=SnowParam(6, progressbar=TRUE))

fitBasic = dream( vobj, ~ CellType + (1|Individual.ID) + Gender.x, info, BPPARAM=SnowParam(6, progressbar=TRUE))

fitDisease = dream( vobj, ~ CellType*Dx + (1|Individual.ID) + Gender.x, info, BPPARAM=SnowParam(6, progressbar=TRUE))
```


### variancePartition
```{r vp}
vp = fitExtractVarPartModel( vobj, ~ (1|CellType) + (1|Dx) + (1|CellType:Dx)+ (1|Individual.ID) + (1|Gender), info, BPPARAM=SnowParam(6, progressbar=TRUE))

plotVarPart( vp )
```




```{r plot.DE.results}

# lapply( levels(info$CellType), function(CT){


#   ensID = rownames(topTable( fit, coef=paste0('CellType', CT), number=10))


#   df = merge( data.frame(Expression = geneExpr[ensID[1],]), info, by="row.names")

#   ggplot(df, aes(CellType, Expression, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5) + scale_color_manual("Cell Type", values=colorCodes) + ggtitle( ensID )




#   } )



```


# Disease by cell type interaction
```{r }

```


















