---
title: "RNA-seq from 4 cell populations"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

cd /hpc/users/hoffmg01/work/cmc_analysis/src
# rm -rf molec_profiling_cache/
ml git pandoc
git pull
R

system("git pull")
rmarkdown::render("molec_profiling.Rmd", output_dir='./', intermediates_dir='./')


# http
# https://hoffmg01.u.hpc.mssm.edu/molec_profiling.html

--->


```{r load.packages, echo=FALSE, message=FALSE, results='hide', cache=FALSE}
suppressPackageStartupMessages({
library(variancePartition)
library(limma)
library(edgeR)
library(mvIC)
library(reshape2)
library(qvalue)
library(data.table)
library(BiocParallel)
library(knitr)
library(org.Hs.eg.db)
library(kableExtra)
library(ggplot2)
library(cowplot)
library(gridExtra)
library(openxlsx)
})

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  cache.lazy = FALSE,  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```

### Reading RNA-seq data
```{r read.data, message=TRUE}
file = "/sc/arion/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/rnaseq/step1-merged/featureCount/featureCount.gene.tsv.gz"
df_featureCounts_gene = fread( file )

geneInfo = df_featureCounts_gene[,1:6]
geneInfo$EnsId = gsub("^(.*)(\\.\\S+)$", "\\1", geneInfo$Geneid)

geneCounts = df_featureCounts_gene[,7:ncol(df_featureCounts_gene)]
geneCounts = as.data.frame(geneCounts)
rownames(geneCounts) = geneInfo$Geneid 

# identify samples with improper names
idx = grep("Undetermined", colnames(geneCounts))
message("Dropping samples with improper names:\n",  paste(colnames(geneCounts)[idx], collapse="\n"))
geneCounts = geneCounts[,-idx]

idx = which(colSums(geneCounts) < 1e5)
message("Dropping samples with zero read counts:\n",  paste(colnames(geneCounts)[idx], collapse="\n"))
geneCounts = geneCounts[,-idx]

# remove genes with duplicated EnsId
exclude = duplicated(geneInfo$EnsId)
geneInfo = geneInfo[!exclude,]
geneCounts = geneCounts[!exclude,]
rownames(geneCounts) = gsub("^(.*)(\\.\\S+)$", "\\1", rownames(geneCounts))

# convert to HUGO names
df_hugo = select(org.Hs.eg.db, geneInfo$EnsId, "SYMBOL", "ENSEMBL")
geneInfo = merge(geneInfo, df_hugo, by.x='EnsId', by.y="ENSEMBL")

```


```{r metadata}
# focused on this experiment
metadata1 = read.xlsx("/sc/arion/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/meta-files/MolecularProfiling_MSSM_120samples_05.24.17-4.xlsx")

# form Synapse
metadataCMC = fread("/sc/arion/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/meta-files/CMC_METADATA.csv")
```

### RNA-seq QC metrics
```{r QC, fig.width=18, fig.height=74}
file = '/sc/arion/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/rnaseq/step1-merged/QCmetrics/qc_metrics.RDS'
qc_metrics.in = t(readRDS( file ))
colnames(qc_metrics.in) = qc_metrics.in[1,]
qc_metrics.in = as.data.frame(qc_metrics.in[-c(1:2),], stringsAsFactors=FALSE)
qc_metrics = apply(qc_metrics.in, 2, as.numeric)
rownames(qc_metrics) = gsub('-', '.', rownames(qc_metrics.in))

# extract information from sample names
info = data.frame( Name = colnames(geneCounts), stringsAsFactors=FALSE)
rownames(info) = info$Name
info$DissectionID = sapply(strsplit( info$Name, '\\.'), function(x) x[1])
info$CellType = sapply(strsplit( info$Name, '\\.'), function(x) x[2])
info$Assay = sapply(strsplit( info$Name, '\\.'), function(x) x[3])

# set color codes cor each cell type
colorCodes = c(
  GABA = '#66A61E',
  GLU = '#E6AB02',
  Olig = '#E7298A',
  MgAs = '#7570B3')

# set factor levels to be the same as the order of the colors
# this makes plotting easier
info$CellType = factor(info$CellType,names(colorCodes))

# add information about each sample
info = merge(info, metadata1, by.x="DissectionID", by.y='Dissection.ID')
info = merge(info, qc_metrics, by.x="Name", by.y='row.names')
rownames(info) = info$Name

# info = merge(info, metadataCMC, by.x='Individual.ID', by.y="Individual ID")
#  info[1:4, 1:20]

# set order to be the same as geneCounts
info = info[match(colnames(geneCounts), info$Name),]

# exclude columns with no variance
index = apply(qc_metrics, 2, function(x) var(x, na.rm=TRUE)) > 0
include = which(index==TRUE)
include = include[! names(include) %in% "PF_BASES"]

df = reshape2::melt(info[,c('DissectionID', 'CellType', colnames(qc_metrics)[include])], id.vars=c('DissectionID', 'CellType'))

ggplot(df, aes( x=CellType, y=value, fill=CellType)) + geom_boxplot() + facet_wrap(~variable, scales="free_y", ncol=5) + theme_bw(16) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + scale_fill_manual("Cell Type", values=colorCodes) 
```




### Study Summary
```{r}
kable(table(info$Dx, info$CellType)) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

```{r}
kable(table(info$Individual.ID)) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

```{r preprocess}
isexpr = rowSums(cpm(geneCounts)>1) >= 0.1*ncol(geneCounts)
table(isexpr)

dge = DGEList( geneCounts[isexpr,] )
dge = calcNormFactors(dge)
```

```{r TMM}
df = merge(dge$samples, info, by="row.names")

ggplot(df, aes(CellType, norm.factors, fill=CellType)) + geom_boxplot() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + ylab("TMM factor")
```

```{r expression.density}

geneExpr = cpm(dge, log=TRUE)

df = reshape2::melt(geneExpr)
df = merge(df, info[,'CellType',drop=FALSE], by.x = "Var2", by.y="row.names")

ggplot(df, aes(value, color=CellType)) + geom_density() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(bquote(log[2]~CPM))

ggplot(df, aes(value, color=Var2)) + geom_density() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + facet_wrap(~CellType, ncol=4) + xlab(bquote(log[2]~CPM))
```




### PCA
```{r PCA}

geneExpr = cpm(dge, log=TRUE)

dcmp = prcomp( t(geneExpr), scale=TRUE)

df_pca = data.frame(dcmp$x)

df_pca = merge(df_pca, info, by.x="row.names", by.y="Name")

varFrac = dcmp$sdev^2 / sum(dcmp$sdev^2)

xlab = paste0('PC1 (', format(varFrac[1]*100, digits=3), '%)')
ylab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
ggplot(df_pca, aes(PC1, PC2, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(xlab) + ylab(ylab)

xlab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
ylab = paste0('PC3 (', format(varFrac[3]*100, digits=3), '%)')
ggplot(df_pca, aes(PC2, PC3, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(xlab) + ylab(ylab)
```      

### Sex check
```{r sex.check}

XIST = 'ENSG00000229807'
UTY = 'ENSG00000183878'

df = data.frame(  Name  = colnames(geneExpr), 
                  XIST  = geneExpr[XIST,],
                  UTY   = geneExpr[UTY,])
df = merge( df, info[,c('Name', 'Gender', 'Individual.ID')], by="Name" )
df$Gender = factor(df$Gender, c("Male", 'Female'))

ggplot(df, aes(XIST, UTY, color=Gender)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=c("blue", "red")) + ggtitle("Sex check")
```

Samples with wrong sex
```{r sex.off}
kable(df[df$Gender == 'Female' & df$UTY > 5,], digits=2, row.names=FALSE) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)  
```



### Esimate precision weights
```{r voom}
form = ~ (1|CellType) + (1|Individual.ID) + (1|Gender)
vobj = voomWithDreamWeights(dge, form, info, plot=TRUE, save.plot=TRUE, BPPARAM=SnowParam(6, progressbar=TRUE))
```

### variancePartition
```{r vp}
form = ~ (1|CellType) + (1|Dx) + (1|CellType:Dx) + (1|Individual.ID) + (1|Gender)
vp = fitExtractVarPartModel( geneExpr, form, info, BPPARAM=SnowParam(6, progressbar=TRUE))

plotVarPart( sortCols(vp) )
```

## Differential expression analysis
### Cell type
```{r DE_process}
form = ~ 0 + CellType + (1|Individual.ID) + (1|Gender)

# Prepare all pairwise contrasts
index = combn(nlevels(info$CellType),2)

L = lapply( 1:ncol(index), function(j){

  CT1 = levels(info$CellType)[index[1,j]]
  CT2 = levels(info$CellType)[index[2,j]]

  L = getContrast( vobj, form, info, paste0('CellType', c(CT1, CT2)))

  L = as.matrix(L)
  colnames(L) = paste0(CT1, '.', CT2)

  L
})
L = do.call(cbind, L)

fitBasic = dream( vobj, form, info, L=L, BPPARAM=SnowParam(6, progressbar=TRUE))
```


```{r DE.analysis}

index = combn(nlevels(info$CellType),2)

df_deCount = lapply( 1:ncol(index), function(j){

    CT1 = levels(info$CellType)[index[1,j]]
    CT2 = levels(info$CellType)[index[2,j]]

    key = paste0(CT1, '.', CT2)

    tabl = topTable(fitBasic, coef=key, number=Inf)

    data.frame(CT1, CT2, key, count = sum(tabl$adj.P.Val < 0.05), pi1 = 1 - qvalue(tabl$P.Value)$pi0)
})
df_deCount = do.call(rbind, df_deCount)

ggplot(df_deCount, aes(CT1, CT2, fill=count)) + geom_tile() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_fill_gradient(low="white", high="red", limits=c(0, 30000)) + geom_text(aes(label=format(count, big.mark=','))) + xlab('') + ylab('') + ggtitle("Count differentiall expression genes")

ggplot(df_deCount, aes(CT1, CT2, fill=pi1)) + geom_tile() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_fill_gradient(name = quote(pi[1]), low="white", high="red", limits=c(0, 1)) + geom_text(aes(label=format(pi1, digits=3))) + xlab('') + ylab('') + ggtitle(bquote(pi[1]))
```

```{r DE2}

index = combn(nlevels(info$CellType),2)

df = lapply( 1:ncol(index), function(j){

    CT1 = levels(info$CellType)[index[1,j]]
    CT2 = levels(info$CellType)[index[2,j]]

    key = paste0(CT1, '.', CT2)

    tabl = topTable(fitBasic, coef=key, number=Inf)

    data.frame(CT1, CT2, key, tabl)
})
df = do.call(rbind, df)

ggplot(df, aes(logFC)) + geom_density(alpha = .2) + facet_wrap(~key) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) 
```

```{r DE.marker_genes, fig.width=15, fig.height=30}

index = combn(nlevels(info$CellType),2)

ensIDs = sapply( 1:ncol(index), function(j){

    CT1 = levels(info$CellType)[index[1,j]]
    CT2 = levels(info$CellType)[index[2,j]]

    key = paste0(CT1, '.', CT2)

    tabl = topTable(fitBasic, coef=key, number=Inf)

    tabl = tabl[order(tabl$t),]

    c(head(rownames(tabl)), tail(rownames(tabl)))
})

ensIDs = unique(c(ensIDs))

df = reshape2::melt(geneExpr[ensIDs,])
df = merge(df, info[,1:3], by.x="Var2", by.y="row.names")

df_hugo = select(org.Hs.eg.db, unique(as.character(df$Var1)), "SYMBOL", "ENSEMBL")
df = merge(df, df_hugo, by.x="Var1", by.y="ENSEMBL")

ggplot(df, aes(CellType, value, fill=CellType)) + geom_boxplot() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position = "none") + scale_fill_manual("Cell Type", values=colorCodes) + facet_wrap(~SYMBOL, ncol=5) + ylab(bquote(log[2]~CPM))
```


# Disease
```{r DE.disease}
# Disease
fitDisease = dream( vobj, ~ 0 + CellType*Dx + (1|Individual.ID) + (1|Gender), info, BPPARAM=SnowParam(6, progressbar=TRUE))
```

```{r Disease.analysis}
df_disease = lapply(levels(info$CellType), function(CT){

  key = paste0('CellType', CT, ':DxSCZ')

  if( CT == levels(info$CellType)[1]){
    key = 'DxSCZ'
  } 

  tabl = topTable( fitDisease, coef = key, number=Inf)

  data.frame(CellType = CT, key, count = sum(tabl$adj.P.Val < 0.05), pi1 = 1 - qvalue(tabl$P.Value)$pi0)
  })
df_disease = do.call(rbind, df_disease)

kable(df_disease, digits=2, row.names=FALSE) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

SOX6

DE from null model
Power vs sample size
Paired versus non-paired
Plot examples of marker genes
distribution of effect size
GTEX brain span.
Compare cell types
Homogeneity of variance for each cell type

For next week: deconvolution of GTEx, PEC, CMC, BrainSpan, BrainVar





### Save mean expression per cell type
https://www.synapse.org/#!Synapse:syn22231629
```{r save.reference.panel}

# compute mean for each cell type, using precision weights
fitCT = dream( vobj, ~ 0 + CellType, info, BPPARAM=SnowParam(6, progressbar=TRUE))
fitCT = eBayes(fitCT)

df_RefPanel = lapply( colnames(fitCT), function(coef){

  # Extract mean for this cell type
  tabl = topTable(fitCT, coef=coef, number=Inf, sort.by='none')
  tabl$se = tabl$logFC / tabl$t # compute standard error given beta and t

  df = data.frame(tabl[,c('logFC', 'se')])
  rownames(df) = rownames(tabl)
  colnames(df) = paste0(gsub('CellType', '', coef), c('', '.se'))
  df

  })
df_RefPanel = do.call(cbind, df_RefPanel)
df_RefPanel = data.frame(EnsID = rownames(df_RefPanel), df_RefPanel)

exclude = grep("\\.se$", colnames(df_RefPanel))

# write mean expression
file = '/hpc/users/hoffmg01/work/cmc_analysis/src/Roussos_MSSM_MolProf_FANS_4_log2CPM_reference_panel.csv.gz'
gz1 <- gzfile(file, "w")
write.table(df_RefPanel[,-exclude], gz1, quote=FALSE, sep=",", row.names=FALSE)
close(gz1)

# write standard error
file = '/hpc/users/hoffmg01/work/cmc_analysis/src/Roussos_MSSM_MolProf_FANS_4_log2CPM_se_reference_panel.csv.gz'
gz1 <- gzfile(file, "w")
write.table(df_RefPanel[,c(1,exclude)], gz1, quote=FALSE, sep=",", row.names=FALSE)
close(gz1)

#  ml python/3.8.2
# synapse add --parentid syn22231629 /hpc/users/hoffmg01/work/cmc_analysis/src/Roussos_MSSM_MolProf_FANS_4_log2CPM_reference_panel.csv.gz
```


```{r exit, cache=FALSE}
knitr::knit_exit()
```



```{r plot.DE.results}

# lapply( levels(info$CellType), function(CT){


#   ensID = rownames(topTable( fit, coef=paste0('CellType', CT), number=10))


#   df = merge( data.frame(Expression = geneExpr[ensID[1],]), info, by="row.names")

#   ggplot(df, aes(CellType, Expression, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5) + scale_color_manual("Cell Type", values=colorCodes) + ggtitle( ensID )




#   } )



```


# Disease by cell type interaction
```{r }

```

# TODO
Genotype concordance
Enrichments
















